airflow/contrib/hooks/gcp_sql_hook.py

```diff
@@ -261,3 +278,541 @@ def _wait_for_operation_to_complete(self, project_id, operation_name):
+        :type project_id: str
+        :param sql_proxy_version: specific version of SQL proxy to download
+            (for example 'v1.13'). By default latest version is downloaded.
+        :type sql_proxy_version: str
+        :param sql_proxy_binary_path: If specified, then proxy will be
+              used from the path specified rather than dynamically generated. This means
+              that if the binary is not present in that path it will also be downloaded.
+        :type sql_proxy_binary_path: str
+        """
+        super(CloudSqlProxyRunner, self).__init__()
+        self.path_prefix = path_prefix
+        if not self.path_prefix:
+            raise AirflowException("The path_prefix must not be empty!")
+        self.sql_proxy_was_downloaded = False
+        self.sql_proxy_version = sql_proxy_version
+        self.download_sql_proxy_dir = None
+        self.sql_proxy_process = None
+        self.instance_specification = instance_specification
+        self.project_id = project_id
+        self.gcp_conn_id = gcp_conn_id
+        self.command_line_parameters = []
+        self.cloud_sql_proxy_socket_directory = self.path_prefix
+        self.sql_proxy_path = sql_proxy_binary_path if sql_proxy_binary_path \
+            else self.path_prefix + "_cloud_sql_proxy"
+        self.credentials_path = self.path_prefix + "_credentials.json"
+        self._build_command_line_parameters()
+
+    def _build_command_line_parameters(self):
+        self.command_line_parameters.extend(
+            ['-dir', self.cloud_sql_proxy_socket_directory])
+        self.command_line_parameters.extend(
+            ['-instances', self.instance_specification])
+
+    @staticmethod
+    def _is_os_64bit():
+        return platform.machine().endswith('64')
+
+    def _download_sql_proxy_if_needed(self):
+        if os.path.isfile(self.sql_proxy_path):
+            self.log.info("cloud-sql-proxy is already present")
+            return
+        system = platform.system().lower()
+        processor = "amd64" if CloudSqlProxyRunner._is_os_64bit() else "386"
+        if not self.sql_proxy_version:
+            download_url = CLOUD_SQL_PROXY_DOWNLOAD_URL.format(system, processor)
+        else:
+            download_url = CLOUD_SQL_PROXY_VERSION_DOWNLOAD_URL.format(
+                self.sql_proxy_version, system, processor)
+        proxy_path_tmp = self.sql_proxy_path + ".tmp"
+        self.log.info("Downloading cloud_sql_proxy from {} to {}".
+                      format(download_url, proxy_path_tmp))
+        r = requests.get(download_url, allow_redirects=True)
+        with open(proxy_path_tmp, 'wb') as f:
+            f.write(r.content)
+        if r.status_code != 200:
+            raise AirflowException(
+                "The cloud-sql-proxy could not be downloaded. Status code = {}. "
+                "Reason = {}".format(r.status_code, r.reason))
+        self.log.info("Moving sql_proxy binary from {} to {}".format(
```

Comment:
It's the traditional way to make sure you have an "atomic" download in case of any concurrent download/execute operations. While you cannot have two writes to the same file ("w" is exclusive) you can easily try to "execute" the binary while it is being downloaded. This can happen if you can have two operators that have the same "fixed" binary path specified and they start almost at the same time. In this case second operator will check if the file is present already and will try to execute it while the first one is still downloading it - which will obviously fail as the binary will be incomplete. OS's move is atomic - either you will get the old or the new file - never partially downloaded ".tmp" file. So in the scenario above both operators will perform download but they will always run "fully downloaded" binary. There is a very slight risk that the first operator cleans up and deletes downloaded file for the second operator in this case but this is far smaller risk as it would have to happen between the check that the binary already exist and running it - which is very, very short comparing to downloading time. Note that even if the file is removed while the second operator runs it, it's no problem - the binary is already running so removing it from the file system will make no difference.

Reply:
I will make a shorter version of the comment in the code.

Reply:
I realised that I should also ignore ENOENT in case i delete downloaded file (to handle cleanly this case on stop_proxy).